compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -3425,7 +3425,7 @@ class Typer extends Namer
       // try converting a constant to the target type
       val folded = ConstFold(tree, pt)
       if (folded ne tree)
-        return adaptConstant(folded, folded.tpe.asInstanceOf[ConstantType])
+        return adaptConstant(tree, folded.tpe.asInstanceOf[ConstantType])
```

Comment:
Hum. Looking at this fix and at `ConstFold.apply(tree, pt)`, I think the latter basically produces an invalid tree, that is never safe to reuse.

I also see that `ConstFold.Apply(tree, pt)` is used *only here*. With the fix, we are therefore patching the invalid result of `ConstFold.apply` at its only call site; but if someone starts calling that overload somewhere else, they will likely reintroduce a similar bug.

Consider inlining the contents of that overload of `ConstFold(tree, tp)` here instead. Once you do that, you won't have to create that transient invalid tree, since you'll be able to directly use `x.convertTo(tp)` instead of `folded.tpe`.

This would prevent further invalid uses of that overload.